aws/resource_aws_route53_resolver_rule_test.go

```diff
@@ -0,0 +1,289 @@
+)
+
+func TestAccAwsRoute53ResolverRule_basic(t *testing.T) {
+	var rule route53resolver.ResolverRule
+	resourceName := "aws_route53_resolver_rule.example"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckRoute53ResolverRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccRoute53ResolverRuleConfig_basicNoTags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRoute53ResolverRuleExists(resourceName, &rule),
+					resource.TestCheckResourceAttr(resourceName, "domain_name", "example.com."),
+					resource.TestCheckResourceAttr(resourceName, "rule_type", "SYSTEM"),
+					resource.TestCheckResourceAttr(resourceName, "share_status", "NOT_SHARED"),
+					testAccCheckResourceAttrAccountID(resourceName, "owner_id"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAwsRoute53ResolverRule_tags(t *testing.T) {
+	var rule route53resolver.ResolverRule
+	resourceName := "aws_route53_resolver_rule.example"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckRoute53ResolverRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccRoute53ResolverRuleConfig_basicTags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRoute53ResolverRuleExists(resourceName, &rule),
+					resource.TestCheckResourceAttr(resourceName, "domain_name", "example.com."),
+					resource.TestCheckResourceAttr(resourceName, "rule_type", "SYSTEM"),
+					resource.TestCheckResourceAttr(resourceName, "share_status", "NOT_SHARED"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "2"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Usage", "original"),
+				),
+			},
+			{
+				Config: testAccRoute53ResolverRuleConfig_basicTagsChanged,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRoute53ResolverRuleExists(resourceName, &rule),
+					resource.TestCheckResourceAttr(resourceName, "domain_name", "example.com."),
+					resource.TestCheckResourceAttr(resourceName, "rule_type", "SYSTEM"),
+					resource.TestCheckResourceAttr(resourceName, "share_status", "NOT_SHARED"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Usage", "changed"),
+				),
+			},
+			{
+				Config: testAccRoute53ResolverRuleConfig_basicNoTags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRoute53ResolverRuleExists(resourceName, &rule),
+					resource.TestCheckResourceAttr(resourceName, "domain_name", "example.com."),
+					resource.TestCheckResourceAttr(resourceName, "rule_type", "SYSTEM"),
+					resource.TestCheckResourceAttr(resourceName, "share_status", "NOT_SHARED"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
```

Comment:
This is a great test case to have in place. My only recommendation is that instead of being here we add it to the basic test check. In fact, it is usually preferred to see default value checks within the basic test validate any unexpected behavior around optional attributes.